src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAssignableTypeForForeachVariable.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    public abstract class UseAssignableTypeForForeachVariable : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2250";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseAssignableTypeForForeachVariableTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseAssignableTypeForForeachVariableMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseAssignableTypeForForeachVariableDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Usage,
+            RuleLevel.BuildWarning,
```

Comment:
@mavasani to confirm but I think you should not set it as build warning.

Reply:
Yes, this should not be a build warning. Only critical API analyzers which are approved by .NET runtime team to be build warnings are allowed to be warnings. This would also be a blocking change.

Reply:
@mavasani Should it be `IdeSuggestion` then? But it says that `IdeSuggestion` is `not something to be enforced in CI`.
This analyzer catches unobvious explicit cast that'll likely result in runtime error. Could you please ping someone from .NET runtime team to verify that it shouldn't be enforced.

Reply:
> But it says that IdeSuggestion is not something to be enforced in CI. This analyzer catches unobvious explicit cast that'll likely result in runtime error. 

@maxkoshevoi IDE Suggestion seems fine. The doc for that field states it is not to be enforced in CI _by default_, end users can bump it to a warning/error if they feel appropriate.

> This analyzer catches unobvious explicit cast that'll likely result in runtime error. 

We only want issues that are critical .NET API issues to be warnings by default. To give more context, if this is a warning, then the next .NET SDK release which contains this analyzer will start breaking everyone's code bases when they just move to a newer SDK version - it is essentially a build breaking change. If you still want to push this as a warning by default, please file a separate issue on dotnet/runtime repo and once you have an approval, we can do the necessary change. Until then, please keep this as an IDE suggestion or Hidden/Bulk configurable diagnostic.

Reply:
@mavasani Changed it to IDE Suggestion for now, and created an issue https://github.com/dotnet/runtime/issues/48529 to discuss if it's appropriate to set it to Warning.

Wouldn't build break only if users have `TreatWarningsAsErrors` set?

Reply:
> Wouldn't build break only if users have TreatWarningsAsErrors set?

Majority of enterprise customers have warnaserror. All warnings shipping in the .NET SDK analyzers are considered build breaking. For example, note that all the CA diagnostics with default severity warning are listed under the build breaking changes in .NET 5.0: https://docs.microsoft.com/dotnet/core/compatibility/5.0#code-analysis

Reply:
Didn't know that. Thanks for the explanation